[ASan][JSON] Unpoison memory before its reuse

AdvenamTacet opened the pull request:
This commit unpoisons memory before its reuse (with reinterpret_cast).
Required by https://github.com/llvm/llvm-project/pull/79049

Notice that it's a temporary solution to prevent buildbots from failing. Read FIXME text for details.

Commit:
[JSON] Unpoison memory before its reuse

This commit unpoisons memory before its reuse (with reinterpret_cast).
Required by https://github.com/llvm/llvm-project/pull/79049

Commit:
clang-format-fix

Commit:
Use LLVM_ADDRESS_SANITIZER_BUILD

aeubanks commented:
could you explain why this is necessary? I'm not understanding at a first glance

vitalybuka commented on llvm/include/llvm/Support/JSON.h:
```diff
@@ -482,6 +482,13 @@ class Value {
   template <typename T, typename... U> void create(U &&... V) {
+#if defined(LLVM_ADDRESS_SANITIZER_BUILD)
+    // Unpoisoning to prevent overwriting poisoned object (e.g., annotated short
+    // string). Objects that have had their memory poisoned may cause an ASan
```
there is a `destroy` method.
could be a bug that this is not called?
I assume std::string will unpoison internal buffer
but if it didn't call ~string, then string with buffer suppose to leak memory?

AdvenamTacet replied:
> there is a `destroy` method.
> could be a bug that this is not called?

`std::basic_string` destructor isn't called. I was looking for a while how I can do it better, but I failed.
Just calling `destroy()` in `create` leads to core dump `free(): invalid pointer`.

> but if it didn't call ~string, then string with buffer suppose to leak memory?
Yes. If that string is a long string, that would lead to leaking memory.
Long term, we definitely should fix that, but it seems to not be easy.

I'm still looking at it and trying to find a way to call a destructor.

PS
Scribbling in ADT may lead to leaking memory as well, but it's only in debug mode.

AdvenamTacet replied:
I didn't find a way to do it yet. Since this stops a larger commit that, to the best of my knowledge, has no bugs, I'm all for upstreaming this temporary solution and then solving the real problem later.

Short string annotations may help with detecting many bugs, therefore I want them to be in LLVM18. The sooner we upstream them, the sooner bugs are detected.

aeubanks replied:
that's fine, but please add a FIXME

vitalybuka replied:
LGTM

AdvenamTacet commented:
> could you explain why this is necessary? I'm not understanding at a first glance

Sure! In https://github.com/llvm/llvm-project/pull/75845 I added annotations for `std::basic_string`, but only when string has an external buffer. Now in https://github.com/llvm/llvm-project/pull/79049#pullrequestreview-1837807790 I'm trying to turn on annotations for short string (so when content is kept inside string object).

In case of a short string, objects memory is poisoned. And if a destructor is not called before `create`, a new object may be created in poisoned memory. (And if string is long, it may lead to a memory leak.)

It would be best to call a destructor, but I didn't figure out how yet and we want to land short string annotations before branching, so it's a workaround to not make buildbots fail.

Commit:
Fix use of LLVM_ADDRESS_SANITIZER_BUILD

This commit fixes incorrect use of `LLVM_ADDRESS_SANITIZER_BUILD` macro.

Details here: https://github.com/llvm/llvm-project/pull/79066#discussion_r1462618308

Commit:
Strange clang-format fix

AdvenamTacet commented:
I ran `buildbot_bootstrap_asan` locally with that change and all ASan tests passed.

Commit:
Add FIXME to the comment.

AdvenamTacet commented:
Buildkite failure is unrelated, therefore I merged.

@vitalybuka @aeubanks Thank you for reviewing and your help! Really appreciate it!
